File: kotlin/vci-client/src/main/java/io/mosip/vciclient/credentialRequest/CredentialRequestFactory.kt
package io.mosip.vciclient.credentialRequest

import io.mosip.vciclient.constants.CredentialFormat
import io.mosip.vciclient.proof.Proof
import io.mosip.vciclient.credentialRequest.types.LdpVcCredentialRequest
import io.mosip.vciclient.credentialRequest.types.MsoMdocCredentialRequest
import io.mosip.vciclient.dto.IssuerMetaData
import io.mosip.vciclient.exception.InvalidDataProvidedException
import okhttp3.Request

class CredentialRequestFactory {
    companion object {
        fun createCredentialRequest(
            credentialFormat: CredentialFormat,
            accessToken: String,
            issuerMetaData: IssuerMetaData,
            proof: Proof,
        ): Request {
            when (credentialFormat) {
                CredentialFormat.LDP_VC -> {
                    return validateAndConstructRequest(
                        LdpVcCredentialRequest(
                            accessToken,
                            issuerMetaData,
                            proof
                        )
                    )
                }

                CredentialFormat.MSO_MDOC -> {
                    return validateAndConstructRequest(
                        MsoMdocCredentialRequest(
                            accessToken,
                            issuerMetaData,
                            proof
                        )
                    )
                }
            }
        }

        private fun validateAndConstructRequest(credentialRequest: CredentialRequest): Request {
            val issuerMetaDataValidatorResult = credentialRequest.validateIssuerMetaData()
            if (issuerMetaDataValidatorResult.isValid)
                return credentialRequest.constructRequest()
            throw InvalidDataProvidedException(issuerMetaDataValidatorResult.invalidFields.toString())
        }
    }
}

File: js/src/credentialRequest/credentialRequestFactory.js
const LdpVcCredentialRequest = require("./types/LdpVcCredentialRequest");
const {CredentialFormat} = require("../constants/CredentialFormat");
const Logger = require("../common/Logger");
const DownloadFailedException = require("../exception/DownloadFailedException");

class CredentialRequestFactory {
   static createCredentialRequest(
    credentialFormat,
    accessToken,
    issuerMetaData,
    proof
  ) {
    switch (credentialFormat) {
      case CredentialFormat.LDP_VC:
        return new LdpVcCredentialRequest(accessToken, issuerMetaData, proof).constructRequest();
      default:
        const errorMessage = `Unsupported credential format type: ${credentialFormat}`;
        Logger.error(errorMessage);
        throw new DownloadFailedException(errorMessage);
    }
  }
}

module.exports = CredentialRequestFactory; 